### PR TITLE
feat: add CCE for gemma3, cohere, and cohere2

### LINKED ISSUE #2438: [ CCE ] Add support for Gemma3 for cut cross entropy
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

There's an upstream PR which has what we want (similar implementation with Gemma2, but with one extra param) https://github.com/apple/ml-cross-entropy/pull/25

### ✔️ Solution

We can wait or patch CCE ourselves incase this is not merged. Waiting for multimodal PR to get in first #2435 

### ❓ Alternatives

_No response_

### 📝 Additional Context

_No response_

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by NanoCode012**
# Description

Add cut cross entropy support for Gemma3 (text and multimodal) and Cohere / Cohere2

Gemma3's implementation was initially adapted from https://github.com/apple/ml-cross-entropy/pull/29 Credit to `@strangedove`.

In gemma3 multimodal, we defer the calculation of logits, so that we can compute within the conditional generation forward pass using LCE.

## How has this been tested?

Gemma3 (text and multimodal) + Cohere2 was tested and vram savings checked.

See https://github.com/axolotl-ai-cloud/axolotl/pull/2443#issuecomment-2749047871 below for sanity benchmarks

### CODE REVIEW

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
+"""Cohere CCE patch."""
```
Comment (salmanmohammadi) [inline src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py]: Does this file need to be attributed?

**Comment by salmanmohammadi** [comment]
LGTM - do you think it's worth adding tests?

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py (hunk 1 above)
Comment (NanoCode012) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py]: Who should we attribute it to?

Comment (salmanmohammadi) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py]: Unsure haha - I just noticed `gemma3.py` had an attribution link but if that's not necessary here that's fine.

Comment (NanoCode012) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py]: Oh right, the upstream PR only had `gemma3` patch.  I thought you meant to Apple.

**Comment by NanoCode012** [comment]
Manually ran some sanity tests for the models involved. Confirmed the patch is applied and training runs.. 

~However, I cannot see much vram savings. Could be due to running qlora + low seq  + batches~

~Edit: Cohere does show vram savings. 13.9GB -> 9.9GB~

Edit2: Will re-run again to be sure.

**Comment by NanoCode012** [comment]
Closes #2438

**Comment by NanoCode012** [comment]
I did small runs with the examples on this branch to get its vram usage before/after. This is done multiple times and we wait a few steps in before taking the result. The value is in the same ballpark each time.

---

Edit: updated table with more models supported


| Model type  | size | Adapter | Before       | After (w CCE) |
| ----------- | ---- | ------- | ------------ | ------------- |
| Gemma       | 2B   | Qlora   | 11.2         | 3.7           |
| Gemma2      | 2B   | Qlora   | 11.7         | 3.9           |
| Gemma3      | 4B   | Lora    | 16 -> 17.8   | 10.4 -> 10.8  |
| Gemma3_text | 1B   | Qlora   | 9.6          | 2.2           |
| Cohere2     | 7B   | Qlora   | 13.9         | 9.8           |
| Mistral     | 7B   | Qlora   | 7.1          | 6.8           |
| Mistral3    | 24B  | Lora    | 39.1         | 37.3          |
| Mllama      | 11B  | Lora    | 23.3         | 22.5          |

Here are the results on a sanity test by passing in some input_ids into the model for before/after loss.

| Model type       | Regular Loss | Patched Loss | Diff < 1e-4                          |
| ---------------- | ------------ | ------------ | ------------------------------------ |
| gemma 2b         | 3.369111     | 3.371979     | N                                    |
| gemma2 2b        | 2.007974     | 2.355560     | N (no softcapping support)           |
| gemma3_text 1b   | 11.755238    | 11.749124    | N                                    |
| gemma3 4b        | 13.435185    | 13.435182    | Y                                    |
| cohere2 7b       | 4.695597     | 4.692424     | N                                    |
| mistral 7b       | 1.463862     | 1.464225     | N                                    |
| mistral3 24B     | 2.812500     | 2.807793     | N (load in 8bit, precision warnings) |
| mllama 11B       | 4.267379     | 4.267569     | N                                    |
|                  |              |              |                                      |
| - original cce - |              |              |                                      |
| llama3 1b        | 6.499690     | 6.498520     | N                                    |

**Comment by NanoCode012** [comment]
> LGTM - do you think it's worth adding tests?

Our current tests mainly are sanity checks. I've done a benchmark above for vram / loss difference. I don't know if it's useful to add as general unit tests.

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py (hunk 1 above)
Comment (winglian) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py]: Maybe add a link to the pr and a TODO to remove it once upstream is merged and published

Comment (NanoCode012) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py]: Will update with changes. 

I found that there is an upstream for cohere right after I made the PR. This one is a bit different as I account for their logit scale.

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+            `input_ids` (special case). Only last token logits are needed for generation, and calculating them only for that
+            token can save memory, which becomes pretty significant for long sequences or large vocabulary size.
+            If a `torch.Tensor`, must be 1D corresponding to the indices to keep in the sequence length dimension.
+            This is useful when using packed tensor format (single dimension for batch and sequence length).
+
+    Returns:
+
+    Example:
+
+    ```python
```
Comment (salmanmohammadi) [inline src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma.py:72]: I don't fully understand how this example is relevant to this function

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma3.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+    >>> generate_ids = model.generate(inputs.input_ids, max_length=30)
+    >>> tokenizer.batch_decode(generate_ids, skip_special_tokens=True, clean_up_tokenization_spaces=False)[0]
+    "What is your favorite condiment?"
+    ```"""
+
+    # pylint: disable=protected-access
+    if self.training and self.config._attn_implementation != "eager":
+        logger.warning_once(
+            "It is strongly recommended to train Gemma3 models with the `eager` attention implementation "
+            f"instead of `{self.config._attn_implementation}`. Use `eager` with `AutoModelForCausalLM.from_pretrained('<path-to-checkpoint>', attn_implementation='eager')`."
```
Comment (salmanmohammadi) [inline src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma3.py]: Should we be suggesting to set `eager_attention` here in the axolotl config?

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma.py (hunk 2 above)
Comment (NanoCode012) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma.py:72]: It's the example doc from the Gemma forward function in transformers https://github.com/huggingface/transformers/blob/2b8a15cc3f1a0c94cf817a8fd8c87bca28737e09/src/transformers/models/gemma/modeling_gemma.py#L801

**Review by salmanmohammadi — APPROVED** [review]
Couple of last nits but LGTM - thank you!

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma3.py (hunk 3 above)
Comment (NanoCode012) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma3.py]: It's the recommendation from Gemma's side. I'm off the opinion to remove this recommendation as it created confusion a few days ago. Should we remove it? I initially kept it to maintain same forward structure as transformers.

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma.py (hunk 2 above)
Comment (salmanmohammadi) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma.py:72]: Gotcha - makes sense since it's a monkeypatch : ) Carry on.

File: src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py (hunk 1 above)
Comment (NanoCode012) [reply src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py]: For gemma3, the upstream PR only works for `gemma3_text` from testing. To work with multimodal, I had to apply the changes listed on the first post.
